hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Server.java

```diff
@@ -2206,7 +2206,25 @@ private void saslProcess(RpcSaslProto saslMessage)
           AUDITLOG.warn(AUTH_FAILED_FOR + this.toString() + ":"
               + attemptingUser + " (" + e.getLocalizedMessage()
               + ") with true cause: (" + tce.getLocalizedMessage() + ")");
-          throw tce;
+          if (!UserGroupInformation.getLoginUser().isLoginSuccess()) {
+            LOG.info("Initiating re-login from IPC Server");
+            if (UserGroupInformation.isLoginKeytabBased()) {
+              UserGroupInformation.getLoginUser().reloginFromKeytab();
```

Comment:
If I trace through the chain of these re-login methods, they end up passing `false` for `ignoreLastLoginTime`. They'll skip the re-login and early exit if insufficient time (default 60 seconds) has elapsed since last login. Would that still leave a server potentially in a bad state for up to 60 seconds?

Reply:
@cnauroth Thanks for review. 

>@surendralilhore , thank you for the patch. I entered a few questions. Additionally, can you please describe if you've been able to simulate the problem in testing to confirm that this patch fixes it? (I assume unit testing isn't practical for this.)

Yes, I simulated it by adding one thread in namenode and that thread will do logout in 2 minute after stating namenode.

Reply:
> Would that still leave a server potentially in a bad state for up to 60 seconds?

Yes, for 60 seconds server will in bad state. Earlier only option was to restart the server. 

Below it the test log for 60 second from my test cluster :

`2022-12-23 10:27:19,117 INFO  ipc.Server - Auth successful for hive/host1-xxxxxxxxxxxx@XXX.SERVER.COM (auth:KERBEROS)
2022-12-23 10:27:19,121 INFO  authorize.ServiceAuthorizationManager - Authorization successful for hive/host1-xxxxxxxxxxxx@XXX.SERVER.COM (auth:KERBEROS) for protocol=interface org.apache.hadoop.hdfs.protocol.ClientProtocol
2022-12-23 10:27:27,048 ERROR namenode.NameNode - Dummy logout thread...
org.apache.hadoop.security.KerberosAuthException: Login failure for user: nn/host1-xxxxxxxxxxxx@XXX.SERVER.COM javax.security.auth.login.LoginException: Re-login failed
        at org.apache.hadoop.security.UserGroupInformation.unprotectedRelogin(UserGroupInformation.java:1203)
        at org.apache.hadoop.hdfs.server.namenode.NameNode$2.run(NameNode.java:1590)
Caused by: javax.security.auth.login.LoginException: Re-login failed
        at org.apache.hadoop.security.UserGroupInformation.unprotectedRelogin(UserGroupInformation.java:1188)
        ... 1 more
2022-12-23 10:27:28,786 WARN  ipc.Server - Auth failed for 10.x.y.z:46879:null (GSS initiate failed) with true cause: (GSS initiate failed)
2022-12-23 10:27:28,786 INFO  ipc.Server - Initiating re-login from IPC Server
2022-12-23 10:27:28,786 INFO  ipc.Server - Doing login from keytab
2022-12-23 10:27:28,786 WARN  security.UserGroupInformation - Not attempting to re-login since the last re-login was attempted less than 60 seconds before. Last Login=1671791247048
.
.
.
.
.
2022-12-23 10:28:27,618 WARN  ipc.Server - Auth failed for 10.x.y.z:45329:null (GSS initiate failed) with true cause: (GSS initiate failed)
2022-12-23 10:28:27,619 INFO  ipc.Server - Initiating re-login from IPC Server
2022-12-23 10:28:27,619 INFO  ipc.Server - Doing login from keytab
2022-12-23 10:28:27,652 INFO  ipc.Server - Retry Auth successful for 10.x.y.z:45329:null after failure
2022-12-23 10:28:27,655 INFO  ipc.Server - Auth successful for hive/hn1-xxxxxxxxxxxx@XXX.SERVER.COM (auth:KERBEROS)
2022-12-23 10:28:27,667 INFO  authorize.ServiceAuthorizationManager - Authorization successful for hive/hn1-xxxxxxxxxxxx@XXX.SERVER.COM (auth:KERBEROS) for protocol=interface org.apache.hadoop.hdfs.protocol.ClientProtocol`

Reply:
@cnauroth , any suggestion to handle this 60 second delay ?

Reply:
For keytab usage, there is `UserGroupInformation#forceReloginFromKeytab()`, which always does the login regardless of time since last login. There is no equivalent `forceReloginFromTicketCache()` though. We could add that, but expanding the public API footprint of `UserGroupInformation` should not be taken lightly. Ideally, I'd like to get a second opinion from one more committer. I think it's the right thing to do. A drawback is that it's potentially a dangerous API if used incorrectly, because it could spam the KDC.

Reply:
Thanks @cnauroth.

Added new API `forceReloginFromTicketCache()` and using both the force API in `Server.java`

>A drawback is that it's potentially a dangerous API if used incorrectly, because it could spam the KDC.

I have added check to use force login API only once in `Server.java` after failure and if it fails again then it will wait for 60 seconds. Handling this by adding **canTryForceLogin** in `Server.java.`

> We could add that, but expanding the public API footprint of UserGroupInformation should not be taken lightly.

Mostly people will use it for new development and they should aware of use case.

> Ideally, I'd like to get a second opinion from one more committer.

@liuml07 Please can you give your opinion here as you reviewed HADOOP-17159

Reply:
Thanks @cnauroth. 

@liuml07 Agree for new force API for re-login.